### DISCUSSION TITLE
Order of onClose calls

### DISCUSSION THREAD

**Discussion by smichovan** [InsertKoinIO/koin#1507] [Q&A]
Is there a way how to change order in witch the onClose methods of components are called?

Lets have two services where the latter is dependent of the former: 

val appModule = module {
    single { Database() } onClose {it?.closeConnectionPool()} 
    single { ServiceUsingDb(get()) } onClose {it?.useDatabaseForCleanup()}
}

Database component maintains a connection to some DB and the second component accesses the DB through the first component. The "ServiceUsingDb" needs to have DB available in onClose method. As the "ServiceUsingDb" is actually dependent on "Database" I would expect Koin to ensure the right closing order. 

I made some research in the Koin code and if I understand it well, the components (instances, mappings) are stored in HashMap, means no specific order is guaranteed. In my case, it may (and does) happen that when ServiceUsingDb onClose is called, the DB is already closed and the cleanup fails. 

Have I missed anything? Is there any way how to influence the onClose methods order?

**Comment by NorbertSandor** [comment]
I have just run into this, maybe somebody knows the solution?

**Comment by arnaudgiuliani** [comment]
Yes, there is effectively no guarantee of order to close components. It could be a new feature. Need to define from where we would like to close things

**Comment by shchuko** [comment]
Please take a look at https://github.com/InsertKoinIO/koin/pull/2020
This fixes the problem in pretty trivial way, hope I did not missed anything

**Comment by simondemartini** [comment]
Is this still planned for 4.3.0? I see [the linked ticket](https://github.com/InsertKoinIO/koin/issues/1510) is closed, and the [the PR](https://github.com/InsertKoinIO/koin/pull/2020) has 4.3.0 as the milestone but is still open.

I think this would be valuable for our work, and I'm happy to take a stab at reviving a PR for this

**Comment by arnaudgiuliani** [comment]
A mistake somewhere 👍 - ticket re-opened
